native/membrane_videocompositor/src/elixir_bridge/elixir_structs.rs

```diff
@@ -136,3 +140,710 @@ impl From<ElixirCropping> for Box<dyn TextureTransformation> {
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_unused_objects(
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        let defined_names = objects.iter().map(|(name, _)| name).collect::<HashSet<_>>();
+
+        let used_names = objects
+            .iter()
+            .flat_map(|(_, object)| object.mentioned_names())
+            .collect::<HashSet<_>>();
+
+        let difference = defined_names.difference(&used_names).collect::<Vec<_>>();
+
+        if difference.is_empty() {
+            return Err(SceneParsingError::CycleDetected);
+        }
+
+        if *difference[0] != final_object_name {
+            return Err(SceneParsingError::UnusedObject((*difference[0]).clone()));
+        }
+
+        if difference.len() > 1 {
+            return Err(SceneParsingError::UnusedObject((*difference[1]).clone()));
+        }
+
+        Ok(())
+    }
+
+    /// returns true if a cycle exists in the scene graph
+    fn contains_cycle(
+        objects: &HashMap<&ObjectName, &Object>,
+        final_object: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        enum NodeState {
+            BeingVisited,
+            Visited,
+        }
+
+        fn visit(
+            name: &ObjectName,
+            objects: &HashMap<&ObjectName, &Object>,
+            visited: &mut HashMap<ObjectName, NodeState>,
+        ) -> Result<(), SceneParsingError> {
+            match visited.get(name) {
+                Some(NodeState::BeingVisited) => return Err(SceneParsingError::CycleDetected),
+                Some(NodeState::Visited) => return Ok(()),
+                _ => {}
+            }
+
+            visited.insert(name.clone(), NodeState::BeingVisited);
+
+            match &objects[name] {
+                Object::Video(_) => Ok(()),
+                Object::Texture(Texture { input, .. }) => visit(&input, objects, visited),
+                Object::Layout(Layout { inputs, .. }) => inputs
+                    .values()
+                    .map(|name| visit(name, objects, visited))
+                    .collect::<Result<Vec<()>, SceneParsingError>>()
+                    .map(|_| ()),
+            }?;
+
+            visited.insert(name.clone(), NodeState::Visited);
+
+            Ok(())
+        }
+
+        let mut visited = HashMap::new();
+
+        visit(final_object, objects, &mut visited)
+    }
+
+    fn validate(&self) -> Result<(), SceneParsingError> {
+        Self::check_for_duplicate_pad_refs(&self.objects)?;
+
+        Self::check_for_duplicate_or_undefined_names(&self.objects, &self.output)?;
+
+        Self::check_for_unused_objects(&self.objects, &self.output)?;
+
+        let objects = self
+            .objects
+            .iter()
+            .map(|(name, object)| (name, object))
+            .collect::<HashMap<_, _>>();
+
+        Self::contains_cycle(&objects, &self.output)?;
+
+        Ok(())
+    }
```

Comment:
I don't think that module named `elixir_bridge/elixir_structs` should be responsible for scene validation.